src/commands/distribute/lib/mc-fus-uploader/mc-fus-uploader.ts

```diff
@@ -0,0 +1,777 @@
+  ambiguousProgress: number = 0;
+  progressUpdateRate: number = 0;
+  maxNumberOfConcurrentUploads: number = 10;
+
+  readonly uploadBaseUrls: any = {
+    CancelUpload: "upload/cancel/",
+    SetMetadata: "upload/set_metadata/",
+    UploadChunk: "upload/upload_chunk/",
+    UploadFinished: "upload/finished/",
+    UploadStatus: "upload/status/",
+  };
+
+  readonly uploadData: IUploadData = {
+    AssetId: "00000000-0000-0000-0000-000000000000",
+    BlobPartitions: 0,
+    CallbackUrl: "",
+    CorrelationId: "00000000-0000-0000-0000-000000000000",
+    CorrelationVector: "",
+    ChunkSize: 0,
+    LogToConsole: false,
+    Tenant: "",
+    UrlEncodedToken: "",
+    TotalBlocks: 0,
+    UploadDomain: "",
+  };
+
+  readonly uploadStatus: IUploadStatus = {
+    AutoRetryCount: 0,
+    AverageSpeed: 0,
+    BlocksCompleted: 0,
+    ChunksFailedCount: 0,
+    ChunkQueue: [],
+    Connected: true,
+    EndTime: new Date(),
+    InflightSet: new Set(),
+    AbortController: new AbortController(),
+    MaxErrorCount: 20,
+    ServiceCallback: {
+      AutoRetryCount: 5,
+      AutoRetryDelay: 1,
+      FailureCount: 0,
+    },
+    StartTime: new Date(),
+    State: McFusUploadState.New,
+    TransferQueueRate: [],
+  };
+
+  readonly eventHandlers: any = {
+    onProgressChanged: (progress: IProgress) => {},
+    onCompleted: (uploadStats: IUploadStats) => {},
+    onResumeRestart: (onResumeStartParams: IOnResumeStartParams) => {},
+    onMessage: (message: string, properties: LogProperties, messageLevel: McFusMessageLevel) => {},
+    onStateChanged: (state: McFusUploadState) => {},
+  };
+
+  constructor(args: IInitializeSettings) {
+    this.initializeUpload(args);
+  }
+
+  calculateAverageSpeed() {
+    if (this.uploadStatus.TransferQueueRate.length === 0) {
+      return 0;
+    }
+
+    let rateSum = 0;
+
+    for (const transferQueueRate of this.uploadStatus.TransferQueueRate) {
+      rateSum += transferQueueRate;
+    }
+
+    return rateSum / this.uploadStatus.TransferQueueRate.length;
+  }
+
+  calculateRate() {
+    // Get the elapsed time in seconds
+    const diff = new Date().getTime() - this.uploadStatus.StartTime.getTime();
+    const seconds = diff / 1000;
+
+    // Megabytes per second
+    const speed = (this.uploadStatus.BlocksCompleted * this.uploadData.ChunkSize) / 1024 / 1024 / seconds;
+
+    // Times 8 to convert bytes to bits
+    const rate = speed * 8;
+    return rate;
+  }
+
+  calculateTimeRemaining() {
+    // calculate time remaining using chunks to avoid hitting the disc for size
+    const dataRemaining = this.uploadStatus.ChunkQueue.length * this.uploadData.ChunkSize;
+    if (this.uploadStatus.AverageSpeed > 0 && dataRemaining > 0) {
+      let timeInSeconds = (dataRemaining * 8) / (1024 * 1024 * this.uploadStatus.AverageSpeed);
+      const hours = Math.floor(timeInSeconds / 60 / 60);
+      timeInSeconds -= hours * 60 * 60;
+      return timeInSeconds;
```

Comment:
This can be simplified with modulo 3600, but the logic here begs the question: what happens if the estimate is over an hour?

Reply:
Addressed in https://github.com/microsoft/appcenter-cli/pull/929/files#diff-6c2f5066e1660e58f6d382c1caa23be1R154